tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+			return
+		}
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) Init(s *Session) {
+	t.getKeyspaceMetadata = s.KeyspaceMetadata
+	go t.decayLoop()
+}
+
+func (t *TokenAwareBanditHostPolicy) Stop() {
+	t.stopOnce.Do(func() {
+		close(t.stopChan)
+	})
+}
+
+// hostInfoSlice converts t.hosts to a slice of *HostInfo.
+// It must be called with t.mu held for reading.
+func (t *TokenAwareBanditHostPolicy) hostInfoSlice() []*HostInfo {
+	hosts := make([]*HostInfo, 0, len(t.hosts))
+	for _, host := range t.hosts {
+		host.mu.RLock()
+		hostInfo := host.hostInfo
+		host.mu.RUnlock()
+		hosts = append(hosts, hostInfo)
+	}
+	return hosts
+}
+
+func (t *TokenAwareBanditHostPolicy) IsLocal(host *HostInfo) bool {
+	return t.localDatacenter != "" && host.DataCenter() == t.localDatacenter
+}
+
+func (t *TokenAwareBanditHostPolicy) KeyspaceChanged(update KeyspaceUpdateEvent) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.updateReplicas(update.Keyspace)
+}
+
+// updateReplicas initiates updating of replicas for the given keyspace.
+// It must be called while the caller holds t.mu mutex locked for writes.
+func (t *TokenAwareBanditHostPolicy) updateReplicas(keyspace string) {
+	ks, ok := t.keyspaces[keyspace]
+	if !ok {
+		ks = &tokenAwareBanditKeyspace{}
+		t.keyspaces[keyspace] = ks
+	}
+
+	ks.needsUpdate = true
+
+	if ks.updating {
+		return
+	}
+
+	ks.updating = true
+	go t.updateReplicasLoop(ks, keyspace)
+}
+
+// replicasNeedUpdate checks if the keyspace needs updating.
+// If the keyspace does not need to be updated, it resets the updating flag.
+func (t *TokenAwareBanditHostPolicy) replicasNeedUpdate(ks *tokenAwareBanditKeyspace) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if ks.needsUpdate {
+		ks.needsUpdate = false
+		return true
+	}
+	ks.updating = false
+	return false
+}
+
+// updateReplicasLoop performs actual update of replicas for a keyspace.
+// updateReplicasLoop must be called in a goroutine not holding the t.mu mutex.
+func (t *TokenAwareBanditHostPolicy) updateReplicasLoop(ks *tokenAwareBanditKeyspace, keyspaceName string) {
+	for t.replicasNeedUpdate(ks) {
+		// We can't call getKeyspaceMetadata while holding the lock since getKeyspaceMetadata can do I/O.
+		metadata, err := t.getKeyspaceMetadata(keyspaceName)
+		if err != nil {
+			t.logger.Println(err)
+			continue
+		}
+
+		strategy := getStrategy(metadata)
+		var replicas tokenRingReplicas
+
+		if strategy != nil {
+			t.mu.RLock()
+			tokenRing := t.tokenRing
+			t.mu.RUnlock()
+			replicas = strategy.replicaMap(tokenRing)
+		}
+
+		t.mu.Lock()
+		ks.replicas = replicas
+		t.mu.Unlock()
+	}
+	if t.afterReplicasUpdated != nil {
+		t.afterReplicasUpdated(keyspaceName)
+	}
+}
+
+// SetPartitioner is called to set the partitioner used.
+func (t *TokenAwareBanditHostPolicy) SetPartitioner(partitioner string) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.partitioner == partitioner {
+		return
+	}
+
+	t.partitioner = partitioner
+
+	err := t.resetRing()
+	if err != nil {
+		Logger.Println(err)
```

Comment:
Are we sure with using of global logger?
```suggestion
		t.logger.Println("cannot reset token ring after setting partitioner:", err)
```

Also shouldn't we switch `t.partitioner` to the old value in case of error?

Reply:
Fixed the logger.

> Also shouldn't we switch t.partitioner to the old value in case of error?

No, the value reflects what the cluster uses, we need to remember the latest one. (`SetPartitioner` is called usually only once during session initialization, since you need to change Cassandra/Scylla config files to change it)

Reply:
Just want to be sure: is it safe to let the wrong/invalid value here and everything will work without a problem?

Reply:
Depends on how you define work without a problem :)

`newTokenRing` parses the partitioner and can fail if the cluster uses a partitioner that is not supported.
If the new partitioner is not supported, we don't update the token ring and use the previous one (if there was any).
If there is no `tokenRing` available, `findReplicas` returns nil and we send queries to the host with lowest latency (from whole cluster).
If there was a tokenRing previously, then something odd happened to the cluster (it's more like you have a completely different cluster, partitioner is not usually changed in an existing cluster, if you change it, you change mapping of which node holds which data). We'll continue to use the old mapping, so if you add nodes to the cluster etc, we won't use them (and the coordinator will resend the queries where they belong). It's really an odd edge case, we can probably set the tokenRing to `nil` in this case so that we consider all nodes when sending queries (i.e. `findReplicas` should return nil)

Reply:
Updated.